MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c

```diff
@@ -0,0 +1,473 @@
+
+      NewNamespaceData = AllocateZeroPool (sizeof (NVME_ADMIN_NAMESPACE_DATA));
+      if (NewNamespaceData == NULL) {
+        Status = EFI_OUT_OF_RESOURCES;
+      } else {
+        Status = NvmeIdentifyNamespace (
+                   Device->Controller,
+                   NamespaceId,
+                   (VOID *)NewNamespaceData
+                   );
+
+        if (!EFI_ERROR (Status)) {
+          //
+          // Update all fields related to LBA size, allocation, and alignment
+          //
+          NewFlbas                = NewNamespaceData->Flbas;
+          LbaFmtIdx               = NewFlbas & NVME_LBA_FORMATNVM_LBAF_MASK;
+          Lbads                   = NewNamespaceData->LbaFormat[LbaFmtIdx].Lbads;
+          Device->Media.BlockSize = (UINT32)1 << Lbads;
+          Device->Media.LastBlock = NewNamespaceData->Nsze - 1;
+
+          CopyMem (&Device->NamespaceData, NewNamespaceData, sizeof (NVME_ADMIN_NAMESPACE_DATA));
+        }
+      }
+    }
+  }
+
+  return Status;
+}
+
+/**
+  Send NVM Express Sanitize Admin Command
+
+  The Sanitize command is used to start a sanitize operation or to recover from a previously
+  failed sanitize operation. The sanitize operation types that may be supported are Block
+  Erase, Crypto Erase, and Overwrite.
+
+  All sanitize operations are processed in the background (i.e., completion of the Sanitize
+  command does not indicate completion of the sanitize operation).
+
+  @param[in] This                Indicates a pointer to the calling context (Block IO Protocol)
+  @param[in] NamespaceId         The NVM Express namespace ID  for which a device path node is to be
+                                 allocated and built. Caller must set the NamespaceId to zero if the
+                                 device path node will contain a valid UUID.
+  @param[in] SanitizeAction      Sanitize action
+  @param[in] NoDeallocAfterSani  No deallocate after sanitize option
+  @param[in] OverwritePatter     Pattern to overwrite old user data
+
+  @retval EFI_SUCCESS            All the namespaces in the device are successfully enumerated.
+  @return Others                 Some error occurs when enumerating the namespaces.
+
+ **/
+EFI_STATUS
+NvmExpressSanitize (
+  IN EFI_BLOCK_IO_PROTOCOL  *This,
+  IN UINT32                 NamespaceId,
+  IN UINT32                 SanitizeAction,
+  IN UINT32                 NoDeallocAfterSanitize,
+  IN UINT32                 OverwritePattern
+  )
+{
+  NVME_DEVICE_PRIVATE_DATA                  *Device;
+  EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET  CommandPacket;
+  EFI_NVM_EXPRESS_COMMAND                   Command;
+  EFI_NVM_EXPRESS_COMPLETION                Completion;
+  NVME_ADMIN_SANITIZE                       SanitizeCdw10Cdw11;
+  EFI_STATUS                                Status;
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+
+  ZeroMem (&CommandPacket, sizeof (EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET));

```

Comment:
I prefer to see sizeof(Command) to keep type errors minimized.  There are pros and cons to this, so I would defer to the "style" from the nvme driver.  You can disregard this comment if the style is to use the type.

Reply:
Understood. I was trying to keep parity with the NVMe driver "style". Below is a sample of the ZeroMem in NvmeIdentifyController() inside NvmExpressHci.c...

  ZeroMem (&CommandPacket, sizeof (EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET));
  ZeroMem (&Command, sizeof (EFI_NVM_EXPRESS_COMMAND));
  ZeroMem (&Completion, sizeof (EFI_NVM_EXPRESS_COMPLETION));

I don't have a preference either way. I'm fine implementing either way. Just let me know which direction to follow.